tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/package.scala

```diff
@@ -1,22 +1,217 @@
 package at.forsyte.apalache.tla
 
 import at.forsyte.apalache.tla.lir._
-import at.forsyte.apalache.tla.lir.oper.TlaOper
+import at.forsyte.apalache.tla.lir.oper._
 import scalaz._
 
 import scala.language.implicitConversions
 
 /**
- * The key definitions related to the typed builder. The most important ones for the users of this API are the methods:
+ * This package defines key types and conversions related to [[ScopedBuilder]].
  *
- *   - an implicit conversion [[typecomp#build]] that converts a builder state to its final form, e.g., to `TlaEx`.
- *   - an implicit conversion [[typecomp#liftBuildToSeq]] that applies `build` to a sequence.
+ * =A brief introduction to [[ScopedBuilder]]=
  *
- * To use the above methods in your code, import the implicit conversions as follows:
+ * [[ScopedBuilder]] is a utility class for generating TLA+ IR expressions. It can be conceptually separated into three
+ * distinct layers:
+ *   1. The signature layer
+ *   1. The type-safe, scope-unsafe layer
+ *   1. The type-safe, scope-safe layer
  *
+ * ==The signature layer==
+ *
+ * Each TLA+ operator has an associated [[TlaOper]] operator in the IR. The majority (but not all) of operators have
+ * type signatures, that is, they restrict the types of the arguments, that may be used to construct valid [[OperEx]]
+ * expressions. For instance, [[TlaArithOper.plus]] represents the arithmetic operator `+` and has the signature `(Int,
+ * Int) => Int` in the type system, meaning that `e @ OperEx(TlaArithOper.plus, x, y)` is considered valid iff `x`, `y`,
+ * and `e` are all tagged with `Typed(IntT1)`. Similarly, [[TlaOper.chooseBounded]] has the signature `(t, Set(t), Bool)
+ * \=> t`, meaning that `e @ OperEx(TlaOper.chooseBounded, x, S, p)` is considered valid if `x` and `e` are tagged with
+ * `Typed(t)`, for some `t`, `S` is tagged with `Typed(SetT1(t))` (for the same `t`), and `p` is tagged with
+ * `Typed(BoolT1)`.
+ *
+ * You can think of a signature `(a1,...,an) => b` as a partial function; given a sequence of argument types `v1, ...
+ * vn`, it returns some type `c` (derived from `b`), if the types `v1, ..., vn` match the pattern `a1, ..., an`. More
+ * precisely, the return value is `s(b)`, if there exists some type substitution `s`, such that `s(ai) = vi` for all `i
+ * \= 1,...,n`.
```

Comment:
Similar here. Maybe we need to adjust `fmt-fix` to be more forgiving to Scaladoc comments.